File: platform/src/retriever.js
// retriever.js
// Graph-only RAG for the assembly graph.
//
// Two stages:
//
//   1. retrieve(question, activeEventId, opts) -> Subgraph
//      Picks a small set of nodes that should ground the answer. Algorithm:
//        a. Seed from the active Event node (+ its phase, component, goal).
//        b. BFS outward to `depth` hops, expanding through any edge type.
//        c. Score each candidate by keyword overlap with the question, then
//           by edge-distance to the seed (closer is better).
//        d. Keep the top-k.
//
//   2. serialize(subgraph) -> { text, refs }
//      Turns the subgraph into a compact text block the LLM can read inline
//      in its system prompt. Each node gets a stable short id (g1, p1, e1,
//      c1, …) so the answer can cite "[e1]" instead of full UUIDs.
//
// The visualization re-uses the SAME subgraph object so the UI shows exactly
// what the model was given. No hidden context.

(function () {
  const STOP = new Set([
    "the","a","an","is","of","to","do","i","do","does","how","what","why","where","which","when","at","on","in","for","and","or","with","without","this","that","my","me","you","be","are","was","were","it","its","but","by","from","as","can","could","should","would","need","want","use","using","step","steps","next","previous"
  ]);
  function tokens(s) {
    return (s || "").toLowerCase().replace(/[^a-z0-9 ]+/g, " ").split(/\s+/).filter((w) => w && !STOP.has(w));
  }

  function score(node, qTokens) {
    if (!qTokens.length) return 0;
    const hay = (node.label + " " + Object.values(node.props || {}).join(" ")).toLowerCase();
    let s = 0;
    qTokens.forEach((t) => { if (hay.includes(t)) s += 1; });
    // Boost when phase/component names exactly match a query token.
    qTokens.forEach((t) => { if (node.label.toLowerCase().split(/[\s_]+/).includes(t)) s += 0.5; });
    return s;
  }

  function bfs(graph, seedIds, maxDepth) {
    const dist = new Map();
    seedIds.forEach((id) => dist.set(id, 0));
    const queue = [...seedIds];
    while (queue.length) {
      const id = queue.shift();
      const d = dist.get(id);
      if (d >= maxDepth) continue;
      graph.neighbors(id).forEach((e) => {
        if (!dist.has(e.other)) {
          dist.set(e.other, d + 1);
          queue.push(e.other);
        }
      });
    }
    return dist;
  }

  function retrieve(graph, question, activeEventId, opts = {}) {
    const { depth = 2, topK = 14 } = opts;
    const qTok = tokens(question);

    // Seeds: active event, its phase, its component, the goal.
    const seeds = new Set();
    const active = graph.get(activeEventId);
    if (active) {
      seeds.add(active.id);
      graph.neighbors(active.id).forEach((e) => {
        const other = graph.get(e.other);
        if (other && (other.kind === "Phase" || other.kind === "Component")) {
          seeds.add(other.id);
        }
      });
    }
    // Always seed the goal so refusals can name it.
    [...graph.nodes.values()].filter((n) => n.kind === "Goal").forEach((g) => seeds.add(g.id));

    const dist = bfs(graph, [...seeds], depth);

    // Rank candidates.
    const scored = [...dist.entries()].map(([id, d]) => {
      const node = graph.get(id);
      const isSeed = seeds.has(id);
      return {
        id, d, isSeed,
        node,
        score: score(node, qTok) * 1.5 + (isSeed ? 1.2 : 0) + Math.max(0, 1.5 - d),
      };
    });
    scored.sort((a, b) => b.score - a.score);
    const keep = new Set(scored.slice(0, topK).map((s) => s.id));
    // Always keep seeds.
    seeds.forEach((id) => keep.add(id));

    // Collect kept nodes + edges that connect them.
    const nodes = [...keep].map((id) => graph.get(id)).filter(Boolean);
    const edges = graph.edges.filter((e) => keep.has(e.from) && keep.has(e.to));

    return { nodes, edges, seeds: [...seeds], activeEventId, question };
  }

  function shortId(node, counts) {
    const prefix = node.kind === "Goal" ? "g" :
                   node.kind === "Phase" ? "p" :
                   node.kind === "Event" ? "e" :
                   node.kind === "Component" ? "c" :
                   node.kind === "Clip" ? "k" : "n";
    counts[prefix] = (counts[prefix] || 0) + 1;
    return prefix + counts[prefix];
  }

  function serialize(subgraph) {
    const counts = {};
    const idMap = new Map();
    const refs = []; // for the UI: short_id -> full_id + node

    subgraph.nodes.forEach((n) => {
      const sid = shortId(n, counts);
      idMap.set(n.id, sid);
      refs.push({ short: sid, node: n });
    });

    const lines = ["NODES:"];
    subgraph.nodes.forEach((n) => {
      const sid = idMap.get(n.id);
      const p = n.props || {};
      let summary = "";
      if (n.kind === "Goal") summary = `target_state=${p.target_state_index} (${p.target_state_name})`;
      else if (n.kind === "Phase") summary = `order=${p.order}, frames=${p.first_frame}-${p.last_frame}, steps=${p.step_count}`;
      else if (n.kind === "Event") summary = `frame=${p.frame}, time=${p.time_s}s, conf=${p.conf}, acts_on=${p.component_name}`;
      else if (n.kind === "Component") summary = `normalized=${p.normalized_name}`;
      else if (n.kind === "Clip") summary = `n_frames=${p.n_frames}, duration=${p.duration_s}s, f1=${p.f1}`;
      lines.push(`  [${sid}] (${n.kind}) "${n.label}" — ${summary}`);
    });

    lines.push("\nEDGES:");
    subgraph.edges.forEach((e) => {
      const a = idMap.get(e.from);
      const b = idMap.get(e.to);
      if (a && b) lines.push(`  [${a}] -[:${e.type}]-> [${b}]`);
    });

    const active = subgraph.nodes.find((n) => n.id === subgraph.activeEventId);
    const activeRef = active ? idMap.get(active.id) : "(none)";

    return {
      text: lines.join("\n"),
      refs,
      idMap,
      activeRef,
    };
  }

  window.Retriever = { retrieve, serialize };
})();

File: backend/retriever.py
"""
Graph-only RAG retrieval.

Mirrors `platform/src/retriever.js` 1:1 so the same inputs produce the same
subgraph regardless of which side runs the retrieval. The frontend and
backend must agree on what the LLM sees — otherwise debugging is hell.
"""
from __future__ import annotations

import re
from collections import deque
from dataclasses import dataclass

from graph_store import Graph
from schemas import GraphNode, Subgraph

_STOPWORDS = {
    "the", "a", "an", "is", "of", "to", "do", "i", "does", "how", "what", "why",
    "where", "which", "when", "at", "on", "in", "for", "and", "or", "with",
    "without", "this", "that", "my", "me", "you", "be", "are", "was", "were",
    "it", "its", "but", "by", "from", "as", "can", "could", "should", "would",
    "need", "want", "use", "using", "step", "steps", "next", "previous",
}


def _tokens(s: str) -> list[str]:
    return [w for w in re.sub(r"[^a-z0-9 ]+", " ", s.lower()).split() if w and w not in _STOPWORDS]


def _node_score(node: GraphNode, q_tokens: list[str]) -> float:
    if not q_tokens:
        return 0.0
    haystack = (node.label + " " + " ".join(str(v) for v in node.props.values())).lower()
    s = sum(1 for t in q_tokens if t in haystack)
    label_words = set(re.split(r"[\s_]+", node.label.lower()))
    s += 0.5 * sum(1 for t in q_tokens if t in label_words)
    return s


def _bfs(graph: Graph, seed_ids: list[str], depth: int) -> dict[str, int]:
    dist: dict[str, int] = {sid: 0 for sid in seed_ids}
    q: deque[str] = deque(seed_ids)
    while q:
        nid = q.popleft()
        d = dist[nid]
        if d >= depth:
            continue
        for _edge, other in graph.neighbors(nid):
            if other not in dist:
                dist[other] = d + 1
                q.append(other)
    return dist


@dataclass
class Retrieved:
    subgraph: Subgraph
    seeds: list[str]


def retrieve(graph: Graph, question: str, active_event_id: str | None,
             depth: int = 2, top_k: int = 14) -> Retrieved:
    seeds: set[str] = set()
    if active_event_id and active_event_id in graph.nodes:
        seeds.add(active_event_id)
        for _e, other in graph.neighbors(active_event_id):
            n = graph.nodes.get(other)
            if n and n.kind in ("Phase", "Component"):
                seeds.add(other)
    for g in graph.by_kind("Goal"):
        seeds.add(g.id)

    if not seeds:
        # Defensive fallback — every clip has a Clip node.
        seeds = {n.id for n in graph.by_kind("Clip")}

    dist = _bfs(graph, list(seeds), depth)
    q_tokens = _tokens(question)

    scored: list[tuple[float, str]] = []
    for nid, d in dist.items():
        node = graph.nodes[nid]
        score = _node_score(node, q_tokens) * 1.5 + (1.2 if nid in seeds else 0) + max(0, 1.5 - d)
        scored.append((score, nid))
    scored.sort(reverse=True)

    keep = {nid for _s, nid in scored[:top_k]}
    keep.update(seeds)

    nodes = [graph.nodes[nid] for nid in keep]
    edges = [e for e in graph.edges if e.src in keep and e.dst in keep]

    return Retrieved(
        subgraph=Subgraph(
            nodes=nodes, edges=edges, seeds=sorted(seeds),
            active_event_id=active_event_id, question=question,
        ),
        seeds=sorted(seeds),
    )


def serialize_for_llm(subgraph: Subgraph) -> tuple[str, dict[str, GraphNode]]:
    """Produce a compact text block + a map from short id ("e2") to the node it
    references. Short ids are what the LLM is told to cite in its answer."""
    counts: dict[str, int] = {}
    id_map: dict[str, str] = {}        # full node id -> short id ("e2")
    ref_map: dict[str, GraphNode] = {}  # short id -> node

    def short_for(node: GraphNode) -> str:
        prefix = {"Goal": "g", "Phase": "p", "Event": "e", "Component": "c", "Clip": "k"}.get(node.kind, "n")
        counts[prefix] = counts.get(prefix, 0) + 1
        return f"{prefix}{counts[prefix]}"

    for n in subgraph.nodes:
        sid = short_for(n)
        id_map[n.id] = sid
        ref_map[sid] = n

    lines = ["NODES:"]
    for n in subgraph.nodes:
        sid = id_map[n.id]
        p = n.props
        if n.kind == "Goal":
            summary = f"target_state={p.get('target_state_index')} ({p.get('target_state_name')})"
        elif n.kind == "Phase":
            summary = f"order={p.get('order')}, key={p.get('phase_key')}"
        elif n.kind == "Event":
            summary = f"frame={p.get('frame')}, time={p.get('time_s')}s, conf={p.get('conf')}, acts_on={p.get('component_name')}"
        elif n.kind == "Component":
            summary = f"normalized={p.get('normalized_name')}"
        elif n.kind == "Clip":
            summary = f"n_frames={p.get('n_frames')}, duration={p.get('duration_s')}s"
        else:
            summary = ""
        lines.append(f"  [{sid}] ({n.kind}) \"{n.label}\" — {summary}")

    lines.append("\nEDGES:")
    for e in subgraph.edges:
        a = id_map.get(e.src); b = id_map.get(e.dst)
        if a and b:
            lines.append(f"  [{a}] -[:{e.type}]-> [{b}]")

    return "\n".join(lines), ref_map


def parse_citations(answer: str) -> list[str]:
    """Return the set of short ids the model cited, in first-occurrence order."""
    seen: set[str] = set()
    out: list[str] = []
    for m in re.finditer(r"\[([a-z]\d+)\]", answer):
        sid = m.group(1)
        if sid not in seen:
            seen.add(sid)
            out.append(sid)
    return out
